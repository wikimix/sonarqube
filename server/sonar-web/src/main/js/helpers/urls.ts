/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { stringify } from 'querystring';
import { omitBy, isNil } from 'lodash';
import {
  isShortLivingBranch,
  isPullRequest,
  isLongLivingBranch,
  getBranchLikeQuery
} from './branches';
import { getProfilePath } from '../apps/quality-profiles/utils';
import { BranchLike, HomePage, HomePageType } from '../app/types';

interface Query {
  [x: string]: string | undefined;
}

interface Location {
  pathname: string;
  query?: Query;
}

export function getBaseUrl(): string {
  return (window as any).baseUrl;
}

export function getHostUrl(): string {
  return window.location.origin + getBaseUrl();
}

export function getPathUrlAsString(path: Location): string {
  return `${getBaseUrl()}${path.pathname}?${stringify(omitBy(path.query, isNil))}`;
}

export function getProjectUrl(project: string): Location {
  return { pathname: '/dashboard', query: { id: project } };
}

export function getPortfolioUrl(key: string): Location {
  return { pathname: '/portfolio', query: { id: key } };
}

export function getComponentBackgroundTaskUrl(componentKey: string, status?: string): Location {
  return { pathname: '/project/background_tasks', query: { id: componentKey, status } };
}

export function getBranchLikeUrl(project: string, branchLike?: BranchLike): Location {
  if (isPullRequest(branchLike)) {
    return getPullRequestUrl(project, branchLike.key);
  } else if (isShortLivingBranch(branchLike)) {
    return getShortLivingBranchUrl(project, branchLike.name);
  } else if (isLongLivingBranch(branchLike)) {
    return getLongLivingBranchUrl(project, branchLike.name);
  } else {
    return getProjectUrl(project);
  }
}

export function getLongLivingBranchUrl(project: string, branch: string): Location {
  return { pathname: '/dashboard', query: { branch, id: project } };
}

export function getShortLivingBranchUrl(project: string, branch: string): Location {
  return { pathname: '/project/issues', query: { branch, id: project, resolved: 'false' } };
}

export function getPullRequestUrl(project: string, pullRequest: string): Location {
  return { pathname: '/project/issues', query: { id: project, pullRequest, resolved: 'false' } };
}

/**
 * Generate URL for a global issues page
 */
export function getIssuesUrl(query: Query, organization?: string): Location {
  const pathname = organization ? `/organizations/${organization}/issues` : '/issues';
  return { pathname, query };
}

/**
 * Generate URL for a component's issues page
 */
export function getComponentIssuesUrl(componentKey: string, query?: Query): Location {
  return { pathname: '/project/issues', query: { ...(query || {}), id: componentKey } };
}

/**
 * Generate URL for a component's drilldown page
 */
export function getComponentDrilldownUrl(
  componentKey: string,
  metric: string,
  branchLike?: BranchLike
): Location {
  return {
    pathname: '/component_measures',
    query: { id: componentKey, metric, ...getBranchLikeQuery(branchLike) }
  };
}

export function getMeasureTreemapUrl(component: string, metric: string) {
  return {
    pathname: '/component_measures',
    query: { id: component, metric, view: 'treemap' }
  };
}

export function getActivityUrl(component: string, branchLike?: BranchLike) {
  return {
    pathname: '/project/activity',
    query: { id: component, ...getBranchLikeQuery(branchLike) }
  };
}

/**
 * Generate URL for a component's measure history
 */
export function getMeasureHistoryUrl(component: string, metric: string, branchLike?: BranchLike) {
  return {
    pathname: '/project/activity',
    query: {
      id: component,
      graph: 'custom',
      custom_metrics: metric,
      ...getBranchLikeQuery(branchLike)
    }
  };
}

/**
 * Generate URL for a component's permissions page
 */
export function getComponentPermissionsUrl(componentKey: string): Location {
  return { pathname: '/project_roles', query: { id: componentKey } };
}

/**
 * Generate URL for a quality profile
 */
export function getQualityProfileUrl(
  name: string,
  language: string,
  organization?: string | null
): Location {
  return getProfilePath(name, language, organization);
}

export function getQualityGateUrl(key: string, organization?: string | null): Location {
  return {
    pathname: getQualityGatesUrl(organization).pathname + '/show/' + encodeURIComponent(key)
  };
}

export function getQualityGatesUrl(organization?: string | null): Location {
  return {
    pathname:
      (organization ? '/organizations/' + encodeURIComponent(organization) : '') + '/quality_gates'
  };
}

/**
 * Generate URL for the rules page
 */
export function getRulesUrl(query: Query, organization: string | null | undefined): Location {
  const pathname = organization ? `/organizations/${organization}/rules` : '/coding_rules';
  return { pathname, query };
}

/**
 * Generate URL for the rules page filtering only active deprecated rules
 */
export function getDeprecatedActiveRulesUrl(
  query: Query = {},
  organization: string | null | undefined
): Location {
  const baseQuery = { activation: 'true', statuses: 'DEPRECATED' };
  return getRulesUrl({ ...query, ...baseQuery }, organization);
}

export function getRuleUrl(rule: string, organization: string | undefined) {
  /* eslint-disable camelcase */
  return getRulesUrl({ open: rule, rule_key: rule }, organization);
  /* eslint-enable camelcase */
}

export function getMarkdownHelpUrl(): string {
  return getBaseUrl() + '/markdown/help';
}

export function getCodeUrl(project: string, branchLike?: BranchLike, selected?: string) {
  return { pathname: '/code', query: { id: project, ...getBranchLikeQuery(branchLike), selected } };
}

export function getOrganizationUrl(organization: string) {
  return `/organizations/${organization}`;
}

export function getHomePageUrl(homepage: HomePage) {
  switch (homepage.type) {
    case HomePageType.Application:
      return getProjectUrl(homepage.component);
    case HomePageType.Project:
      return homepage.branch
        ? getLongLivingBranchUrl(homepage.component, homepage.branch)
        : getProjectUrl(homepage.component);
    case HomePageType.Organization:
      return getOrganizationUrl(homepage.organization);
    case HomePageType.Portfolio:
      return getPortfolioUrl(homepage.component);
    case HomePageType.Portfolios:
      return '/portfolios';
    case HomePageType.MyProjects:
      return '/projects';
    case HomePageType.Issues:
    case HomePageType.MyIssues:
      return { pathname: '/issues', query: { resolved: 'false' } };
  }

  // should never happen, but just in case...
  return '/projects';
}

export function isUrl(url: string) {
  if (!URL) {
    const elem = document.createElement('a');
    elem.href = url;
    return !!(elem.host && elem.hostname && elem.protocol);
  }
  try {
    const parsedUrl = new URL(url);
    return url.includes(parsedUrl.host);
  } catch (error) {
    return false;
  }
}
