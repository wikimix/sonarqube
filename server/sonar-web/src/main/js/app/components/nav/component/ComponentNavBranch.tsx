/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import * as PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import ComponentNavBranchesMenu from './ComponentNavBranchesMenu';
import SingleBranchHelperPopup from './SingleBranchHelperPopup';
import NoBranchSupportPopup from './NoBranchSupportPopup';
import { BranchLike, Component } from '../../../types';
import * as theme from '../../../theme';
import BranchIcon from '../../../../components/icons-components/BranchIcon';
import {
  isShortLivingBranch,
  isSameBranchLike,
  getBranchLikeDisplayName,
  isPullRequest
} from '../../../../helpers/branches';
import { translate } from '../../../../helpers/l10n';
import HelpIcon from '../../../../components/icons-components/HelpIcon';
import BubblePopupHelper from '../../../../components/common/BubblePopupHelper';
import Tooltip from '../../../../components/controls/Tooltip';

interface Props {
  branchLikes: BranchLike[];
  component: Component;
  currentBranchLike: BranchLike;
  location?: any;
}

interface State {
  dropdownOpen: boolean;
  noBranchSupportPopupOpen: boolean;
  singleBranchPopupOpen: boolean;
}

export default class ComponentNavBranch extends React.PureComponent<Props, State> {
  mounted = false;

  static contextTypes = {
    branchesEnabled: PropTypes.bool.isRequired,
    onSonarCloud: PropTypes.bool
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      dropdownOpen: false,
      noBranchSupportPopupOpen: false,
      singleBranchPopupOpen: false
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillReceiveProps(nextProps: Props) {
    if (
      nextProps.component !== this.props.component ||
      !isSameBranchLike(nextProps.currentBranchLike, this.props.currentBranchLike) ||
      nextProps.location !== this.props.location
    ) {
      this.setState({ dropdownOpen: false, singleBranchPopupOpen: false });
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleClick = (event: React.SyntheticEvent<HTMLElement>) => {
    event.preventDefault();
    event.stopPropagation();
    event.currentTarget.blur();
    this.setState(state => ({ dropdownOpen: !state.dropdownOpen }));
  };

  closeDropdown = () => {
    if (this.mounted) {
      this.setState({ dropdownOpen: false });
    }
  };

  toggleSingleBranchPopup = (show?: boolean) => {
    if (show !== undefined) {
      this.setState({ singleBranchPopupOpen: show });
    } else {
      this.setState(state => ({ singleBranchPopupOpen: !state.singleBranchPopupOpen }));
    }
  };

  toggleNoBranchSupportPopup = (show?: boolean) => {
    if (show !== undefined) {
      this.setState({ noBranchSupportPopupOpen: show });
    } else {
      this.setState(state => ({ noBranchSupportPopupOpen: !state.noBranchSupportPopupOpen }));
    }
  };

  handleSingleBranchClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.stopPropagation();
    this.toggleSingleBranchPopup();
  };

  handleNoBranchSupportClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.stopPropagation();
    this.toggleNoBranchSupportPopup();
  };

  renderDropdown = () => {
    const { configuration } = this.props.component;
    return this.state.dropdownOpen ? (
      <ComponentNavBranchesMenu
        branchLikes={this.props.branchLikes}
        canAdmin={configuration && configuration.showSettings}
        component={this.props.component}
        currentBranchLike={this.props.currentBranchLike}
        onClose={this.closeDropdown}
      />
    ) : null;
  };

  renderMergeBranch = () => {
    const { currentBranchLike } = this.props;
    if (isShortLivingBranch(currentBranchLike)) {
      return currentBranchLike.isOrphan ? (
        <span className="note big-spacer-left text-lowercase">
          {translate('branches.orphan_branch')}
          <Tooltip overlay={translate('branches.orphan_branches.tooltip')}>
            <i className="icon-help spacer-left" />
          </Tooltip>
        </span>
      ) : (
        <span className="note big-spacer-left text-lowercase">
          {translate('from')} <strong>{currentBranchLike.mergeBranch}</strong>
        </span>
      );
    } else if (isPullRequest(currentBranchLike)) {
      return (
        <span className="note big-spacer-left text-lowercase">
          <FormattedMessage
            defaultMessage={translate('branches.pull_request.for_merge_into_x_from_y')}
            id="branches.pull_request.for_merge_into_x_from_y"
            values={{
              base: <strong>{currentBranchLike.base}</strong>,
              branch: <strong>{currentBranchLike.branch}</strong>
            }}
          />
        </span>
      );
    } else {
      return null;
    }
  };

  renderSingleBranchPopup = () => (
    <div className="display-inline-block spacer-left">
      <a className="link-no-underline" href="#" onClick={this.handleSingleBranchClick}>
        <HelpIcon fill={theme.blue} />
      </a>
      <BubblePopupHelper
        isOpen={this.state.singleBranchPopupOpen}
        position="bottomleft"
        popup={<SingleBranchHelperPopup />}
        togglePopup={this.toggleSingleBranchPopup}
      />
    </div>
  );

  renderNoBranchSupportPopup = () => (
    <div className="display-inline-block spacer-left">
      <a className="link-no-underline" href="#" onClick={this.handleNoBranchSupportClick}>
        <HelpIcon fill={theme.gray80} />
      </a>
      <BubblePopupHelper
        isOpen={this.state.noBranchSupportPopupOpen}
        position="bottomleft"
        popup={<NoBranchSupportPopup />}
        togglePopup={this.toggleNoBranchSupportPopup}
      />
    </div>
  );

  render() {
    const { branchLikes, currentBranchLike } = this.props;

    if (this.context.onSonarCloud && !this.context.branchesEnabled) {
      return null;
    }

    const displayName = getBranchLikeDisplayName(currentBranchLike);

    if (!this.context.branchesEnabled) {
      return (
        <div className="navbar-context-branches">
          <BranchIcon
            branchLike={currentBranchLike}
            className="little-spacer-right"
            fill={theme.gray80}
          />
          <span className="note">{displayName}</span>
          {this.renderNoBranchSupportPopup()}
        </div>
      );
    }

    if (branchLikes.length < 2) {
      return (
        <div className="navbar-context-branches">
          <BranchIcon branchLike={currentBranchLike} className="little-spacer-right" />
          <span className="note">{displayName}</span>
          {this.renderSingleBranchPopup()}
        </div>
      );
    }

    return (
      <div
        className={classNames('navbar-context-branches', 'dropdown', {
          open: this.state.dropdownOpen
        })}>
        <a className="link-base-color link-no-underline" href="#" onClick={this.handleClick}>
          <BranchIcon branchLike={currentBranchLike} className="little-spacer-right" />
          <Tooltip overlay={displayName} mouseEnterDelay={1}>
            <span className="text-limited text-top">{displayName}</span>
          </Tooltip>
          <i className="icon-dropdown little-spacer-left" />
        </a>
        {this.renderDropdown()}
        {this.renderMergeBranch()}
      </div>
    );
  }
}
