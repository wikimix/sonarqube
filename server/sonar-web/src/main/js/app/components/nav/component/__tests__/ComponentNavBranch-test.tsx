/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ComponentNavBranch from '../ComponentNavBranch';
import {
  BranchType,
  ShortLivingBranch,
  MainBranch,
  Component,
  LongLivingBranch,
  PullRequest
} from '../../../../types';
import { click } from '../../../../../helpers/testUtils';

const mainBranch: MainBranch = { isMain: true, name: 'master' };
const fooBranch: LongLivingBranch = { isMain: false, name: 'foo', type: BranchType.LONG };

it('renders main branch', () => {
  const component = {} as Component;
  expect(
    shallow(
      <ComponentNavBranch
        branchLikes={[mainBranch, fooBranch]}
        component={component}
        currentBranchLike={mainBranch}
      />,
      { context: { branchesEnabled: true } }
    )
  ).toMatchSnapshot();
});

it('renders short-living branch', () => {
  const branch: ShortLivingBranch = {
    isMain: false,
    mergeBranch: 'master',
    name: 'foo',
    status: { bugs: 0, codeSmells: 0, vulnerabilities: 0 },
    type: BranchType.SHORT
  };
  const component = {} as Component;
  expect(
    shallow(
      <ComponentNavBranch
        branchLikes={[branch, fooBranch]}
        component={component}
        currentBranchLike={branch}
      />,
      { context: { branchesEnabled: true } }
    )
  ).toMatchSnapshot();
});

it('renders pull request', () => {
  const pullRequest: PullRequest = {
    base: 'master',
    branch: 'feature',
    key: '1234',
    title: 'Feature PR',
    url: 'https://example.com/pull/1234'
  };
  const component = {} as Component;
  expect(
    shallow(
      <ComponentNavBranch
        branchLikes={[pullRequest, fooBranch]}
        component={component}
        currentBranchLike={pullRequest}
      />,
      { context: { branchesEnabled: true } }
    )
  ).toMatchSnapshot();
});

it('opens menu', () => {
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch
      branchLikes={[mainBranch, fooBranch]}
      component={component}
      currentBranchLike={mainBranch}
    />,
    { context: { branchesEnabled: true } }
  );
  expect(wrapper.find('ComponentNavBranchesMenu')).toHaveLength(0);
  click(wrapper.find('a'));
  expect(wrapper.find('ComponentNavBranchesMenu')).toHaveLength(1);
});

it('renders single branch popup', () => {
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch
      branchLikes={[mainBranch]}
      component={component}
      currentBranchLike={mainBranch}
    />,
    { context: { branchesEnabled: true } }
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(false);
  click(wrapper.find('a'));
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(true);
});

it('renders no branch support popup', () => {
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch
      branchLikes={[mainBranch, fooBranch]}
      component={component}
      currentBranchLike={mainBranch}
    />,
    { context: { branchesEnabled: false } }
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(false);
  click(wrapper.find('a'));
  expect(wrapper.find('BubblePopupHelper').prop('isOpen')).toBe(true);
});

it('renders nothing on SonarCloud without branch support', () => {
  const component = {} as Component;
  const wrapper = shallow(
    <ComponentNavBranch
      branchLikes={[mainBranch]}
      component={component}
      currentBranchLike={mainBranch}
    />,
    { context: { branchesEnabled: false, onSonarCloud: true } }
  );
  expect(wrapper.type()).toBeNull();
});
