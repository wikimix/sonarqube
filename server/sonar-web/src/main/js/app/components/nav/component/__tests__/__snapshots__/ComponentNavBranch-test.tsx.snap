// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders main branch 1`] = `
<div
  className="navbar-context-branches dropdown"
>
  <a
    className="link-base-color link-no-underline"
    href="#"
    onClick={[Function]}
  >
    <BranchIcon
      branchLike={
        Object {
          "isMain": true,
          "name": "master",
        }
      }
      className="little-spacer-right"
    />
    <Tooltip
      mouseEnterDelay={1}
      overlay="master"
      placement="bottom"
    >
      <span
        className="text-limited text-top"
      >
        master
      </span>
    </Tooltip>
    <i
      className="icon-dropdown little-spacer-left"
    />
  </a>
</div>
`;

exports[`renders no branch support popup 1`] = `
<div
  className="navbar-context-branches"
>
  <BranchIcon
    branchLike={
      Object {
        "isMain": true,
        "name": "master",
      }
    }
    className="little-spacer-right"
    fill="#cdcdcd"
  />
  <span
    className="note"
  >
    master
  </span>
  <div
    className="display-inline-block spacer-left"
  >
    <a
      className="link-no-underline"
      href="#"
      onClick={[Function]}
    >
      <HelpIcon
        fill="#cdcdcd"
      />
    </a>
    <BubblePopupHelper
      isOpen={false}
      popup={<NoBranchSupportPopup />}
      position="bottomleft"
      togglePopup={[Function]}
    />
  </div>
</div>
`;

exports[`renders pull request 1`] = `
<div
  className="navbar-context-branches dropdown"
>
  <a
    className="link-base-color link-no-underline"
    href="#"
    onClick={[Function]}
  >
    <BranchIcon
      branchLike={
        Object {
          "base": "master",
          "branch": "feature",
          "key": "1234",
          "title": "Feature PR",
          "url": "https://example.com/pull/1234",
        }
      }
      className="little-spacer-right"
    />
    <Tooltip
      mouseEnterDelay={1}
      overlay="1234 – Feature PR"
      placement="bottom"
    >
      <span
        className="text-limited text-top"
      >
        1234 – Feature PR
      </span>
    </Tooltip>
    <i
      className="icon-dropdown little-spacer-left"
    />
  </a>
  <span
    className="note big-spacer-left text-lowercase"
  >
    <FormattedMessage
      defaultMessage="branches.pull_request.for_merge_into_x_from_y"
      id="branches.pull_request.for_merge_into_x_from_y"
      values={
        Object {
          "base": <strong>
            master
          </strong>,
          "branch": <strong>
            feature
          </strong>,
        }
      }
    />
  </span>
</div>
`;

exports[`renders short-living branch 1`] = `
<div
  className="navbar-context-branches dropdown"
>
  <a
    className="link-base-color link-no-underline"
    href="#"
    onClick={[Function]}
  >
    <BranchIcon
      branchLike={
        Object {
          "isMain": false,
          "mergeBranch": "master",
          "name": "foo",
          "status": Object {
            "bugs": 0,
            "codeSmells": 0,
            "vulnerabilities": 0,
          },
          "type": "SHORT",
        }
      }
      className="little-spacer-right"
    />
    <Tooltip
      mouseEnterDelay={1}
      overlay="foo"
      placement="bottom"
    >
      <span
        className="text-limited text-top"
      >
        foo
      </span>
    </Tooltip>
    <i
      className="icon-dropdown little-spacer-left"
    />
  </a>
  <span
    className="note big-spacer-left text-lowercase"
  >
    from
     
    <strong>
      master
    </strong>
  </span>
</div>
`;

exports[`renders single branch popup 1`] = `
<div
  className="navbar-context-branches"
>
  <BranchIcon
    branchLike={
      Object {
        "isMain": true,
        "name": "master",
      }
    }
    className="little-spacer-right"
  />
  <span
    className="note"
  >
    master
  </span>
  <div
    className="display-inline-block spacer-left"
  >
    <a
      className="link-no-underline"
      href="#"
      onClick={[Function]}
    >
      <HelpIcon
        fill="#4b9fd5"
      />
    </a>
    <BubblePopupHelper
      isOpen={false}
      popup={<SingleBranchHelperPopup />}
      position="bottomleft"
      togglePopup={[Function]}
    />
  </div>
</div>
`;
